src/Unsafe/Linear.hs

```diff
@@ -29,8 +32,11 @@ import qualified Unsafe.Coerce as NonLinear
 import GHC.Exts (TYPE, RuntimeRep)
 
 -- | Linearly typed @unsafeCoerce@
-coerce :: a %1-> b
-coerce = NonLinear.unsafeCoerce NonLinear.unsafeCoerce
+coerce :: forall a b. a %1-> b
+coerce a =
+  case NonLinear.unsafeEqualityProof @a @b of
+    NonLinear.UnsafeRefl -> a
+{-# INLINE coerce #-}
```

Comment:
This `INLINE` actually does not have an effect on my test case, since GHC already was inlining it even without. But I thought it's better to be explicit.